File: challenge_6/cpp/manuel/src/ranges.cpp
#include <vector>
#include <string>

namespace ranges {

    std::string represent_range (int start, int end) {

        return std::to_string(start) + "->" + std::to_string(end);
    }

    std::vector<std::string> get_ranges (std::vector<int> numbers) {
        // [1, 2, 4, 5] to ["1->2", "4->5"]

        std::vector<std::string> ranges;

        int length = numbers.size();
        int start = numbers[0];
        int end = numbers[0];

        for( int i = 1; i < length; i++ ) {

            // If there's a gap of >=2 and the current range is > 0
            if( end + 1 < numbers[i] && start != end) {

                ranges.push_back(represent_range(start, end));
                start = numbers[i];
            } else if (end + 1 < numbers[i] && start == end) {
                // if the gap is >=2, but the range is 0
                // skip

                start = numbers[i];
            }

            end = numbers[i];
        }

        // Final range
        if (start != end) {
            ranges.push_back(represent_range(start, end));
        }

        return ranges;
    }

}

File: challenge_6/python/bryantpq/ranges.py
def get_ranges(a):
    result = []

    if a:
        start = a[0]
        found = False

        for i in range(1, len(a)):
            if a[i] - 1 != a[i - 1]:
                if found:
                    result.append(str(start) + "->" + str(a[i - 1]))
                    found = False
                start = a[i]
            else:
                found = True

        if found:
           result.append(str(start) + "->" + str(a[i]))
 

    return result


def main():
    a = [1, 2, 4, 6, 7, 8, 10, 20, 21, 25, 27, 29, 30, 31, 32]
    print(get_ranges(a))
    b = [1, 2, 3, 4, 8, 9, 10, 12, 13, 14]
    print(get_ranges(b))
    c = [1]
    print(get_ranges(c))
    d = []
    print(get_ranges(d))

if __name__ == "__main__":
    main()
